src/server/pfs/server/driver.go

```diff
@@ -3310,28 +3285,76 @@ func (s *putFileServer) Peek() (*pfs.PutFileRequest, error) {
 	return req, nil
 }
 
-func forEachPutFile(server pfs.API_PutFileServer, f func(*pfs.PutFileRequest, io.Reader) error) error {
+func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_PutFileServer, f func(*pfs.PutFileRequest, io.Reader) error) (bool, string, string, error) {
 	limiter := limit.New(client.DefaultMaxConcurrentStreams)
 	var pr *io.PipeReader
 	var pw *io.PipeWriter
 	var req *pfs.PutFileRequest
 	var err error
 	var eg errgroup.Group
+	oneOff := false
+	var rawCommitID string
+	var commitID string
+	var repo string
+	var branch string
+
 	for req, err = server.Recv(); err == nil; req, err = server.Recv() {
 		req := req
 		if req.File != nil {
+			// For the first file, dereference the commit ID if needed. For
+			// subsequent files, ensure that they're all referencing the same
+			// commit ID, and replace their commit objects to all refer to the
+			// same thing.
+			if rawCommitID == "" {
+				commit := req.File.Commit
+				repo = commit.Repo.Name
+				// The non-dereferenced commit ID. Used to ensure that all
+				// subsequent requests use the same value.
+				rawCommitID = commit.ID
+				// oneOff is true if we're creating the commit as part of this put-file
+				oneOff = false
+				// inspectCommit will replace file.Commit.ID with an actual commit ID if
+				// it's a branch. So we want to save it first.
+				if !uuid.IsUUIDWithoutDashes(commit.ID) {
+					branch = commit.ID
+				}
+				commitInfo, err := d.inspectCommit(pachClient, commit, pfs.CommitState_STARTED)
```

Comment:
Would you be on board with adding a comment here like:
```
// inspect the commit where we're adding files and figure out if this
// is a one-off put-file. 
// - if 'commit' refers to an open commit                -> not oneOff
// - otherwise (i.e. branch with closed HEAD or no HEAD) -> yes oneOff
// Note that if commit is a specific commit ID, it must be open for
// this call to succeed
```
Just because the code here is doing a lot, but I think the question it's answering is pretty short